iceoryx_binding_c/include/iceoryx_binding_c/types.h

```diff
@@ -15,34 +15,39 @@
 #ifndef IOX_BINDING_C_TYPES_H
 #define IOX_BINDING_C_TYPES_H
 
-/// @brief describes the current state of a subscriber
-enum iox_SubscribeState
+#include "internal/c2cpp_binding.h"
+
+struct iox_wait_set_storage_t_
+{
+    // the value of the array size is the result of the following formula:
+    // sizeof(WaitSet) / 8
+    uint64_t do_not_touch_me[133];
```

Comment:
Why hard coding all the values instead of calculating them with the `sizeof` operator? Then you can remove the comments.

Reply:
This is not possible. To use the sizeof operator I need to know the class which is impossible from C. This header is included by C and C++ and since C has no concept of classes I have to use a compile time expression which C understands.

The only thing left is a macro or I write the number directly into the struct. I decided for a magic number with a comment.

Reply:
this is getting more and more messy. From one magic number we went to 5. We have to think of a more viable solution. Maybe some cmake magic can help, but I also don't know a way to do it.

Reply:
@elBoberido this is correct it is kind of a controlled mess. We have unit tests in place which are verifying those number (see iceoryx_binding_c/test/integrationtests/test_types.cpp).

I somehow doubt that CMake magic can help here. The reason is that CMake then has to parse C++ code, execute sizeof(MyClass). We cannot write a static number into cmake since these structs have different sizes on different platforms (see `iox_guard_cond_storage_t`).

And I think in the future we will get even more magic storage structs in here. But at the moment my approach is the following:
1. Add new struct with comment.
2. Tests the size and alignment in the `test_types.cpp` unit tests which detects reliable when something is off (with this approach I discovered that the `iox_guard_cond_storage_t` is different on mac)

Reply:
I am wondering if the best solution would be a minimalistic C++ code generator which generates exactly this header. 
We could realize it like we did with TOML. When you call `cmake -Bbuild -H.....` we compile this minimalistic C++ code generator and generate this file?!
What do you think @elBoberido  @mossmaurice @marthtz

Reply:
These hardcoded values are really bad when someone adapts the classes (potentially changing the size of e.g. Waitset) but not the sizes in the header here...from the top of my head I do not have a solution yet.

Reply:
I'm voting for some sort of generator here. Maybe we can compile just e.g the `WaitSet` class beforehand, call `sizeof(WaitSet)` and then generate this file with CMake?

Reply:
To reduce the risk a bit one could add the size to the storage class and check it at runtime when the object is initialized there. Of course this incurs runtime overhead and the size still needs to be adapted when the size of the underlying object changes. But it would fail then and there and not potentially corrupting memory and fail later.
If available we need a better solution, I just do not have one right now.

Reply:
Getting quite complex here. All these magic numbers are not that cool, we all agree on that. How easy would it be to have something to create a file to include with all the necessary sizes? Add stuff to runtime is not desirable I guess.

Reply:
I would propose the following solution. For now we leave it as it is for now. I have unit tests which verify that the size is correct so we have a safety net in place!
Right after the c binding is merged (and at the moment the c binding is in a experimental state) I will explore this topic further and try to find a suitable solution.

If I try to follow up on another approach this pull request would explode. If I would try to implement some kind of C++ to C code generator we would get another 20 or 30 files in here and it would take at least four more weeks.

Reply:
@budrus @dkroenke your opinions or ideas are missing. Can one of you pull a bunny out of the head which creates these structs at compile time for us by using somehow magically `sizeof(CppClass)` directly from C.

Reply:
Ok to make the problem clear: You want to have at compile time the size of class WaitSet which is a C++ Class.
The only thing which came to my mind is maybe to use <extern "C++">  so that the corresponding linker is called. 
But for this you need to include then the Waitset header here too.

Reply:
@dkroenke yes, in order to be able to use `sizeof` we would need to include the C++ header. This would break the C compilation. The usual way to solve this is by using the heap and returning a pointer to the data. But we cannot do this, so we need to put it on the stack. On the stack we need to know the size at compile time, therefore this magic numbers. It's not nice and we should look for a better solution, but it seems to be the best of only bad solutions and we have tests to check for breakage.

Reply:
@elfenpiff please leave a comment on the array with the iox-ticket https://github.com/eclipse/iceoryx/issues/308 
to track this.